src/main/java/org/sonatype/nexus/ci/nxrm/v3/ComponentUploaderNxrm3.groovy

```
@@ -0,0 +1,94 @@
+
+
+        def mavenComponentBuilder = MavenComponentBuilder.create()
+            .withGroupId(envVars.expand(mavenCoordinate.groupId))
+            .withArtifactId(envVars.expand(mavenCoordinate.artifactId))
+            .withVersion(envVars.expand(mavenCoordinate.version))
+            .withPackaging(envVars.expand(mavenCoordinate.packaging))
+
+        remoteMavenAssets.eachWithIndex { remoteMavenAsset, idx ->
+          def asset = new DefaultAsset("asset${idx}", remoteMavenAsset.RemotePath.getRemote(),
+              remoteMavenAsset.RemotePath.read())
+
+          mavenComponentBuilder.withAsset(asset, envVars.expand(remoteMavenAsset.Asset.extension),
+              envVars.expand(remoteMavenAsset.Asset.classifier))
+        }
+
+        try {
+          nxrmClient.upload(nxrmRepositoryId, mavenComponentBuilder.build())
+        }
+        catch (RepositoryManagerException ex) {
+          throw new IOException(ex)
+        }
+      }
+      catch (IOException ex) {
+        final String uploadFailed = 'Upload of maven component with GAV ' +
+            "[${mavenCoordinate.groupId}:${mavenCoordinate.artifactId}:${mavenCoordinate.version}] failed"
+
+        logger.println(uploadFailed)
+        logger.println('Failing build due to failure to upload file to Nexus Repository Manager Publisher')
+        run.setResult(Result.FAILURE)
+        throw new IOException(uploadFailed, ex)
+      }
+    }
+
+    logger.println('Successfully Uploaded Maven Assets')
+  }
+
+  RepositoryManagerV3Client getRepositoryManagerClient(final Nxrm3Configuration nxrm3Configuration) {
+    try {
+      nexus3Client(nxrm3Configuration.serverUrl, nxrm3Configuration.credentialsId, nxrm3Configuration.anonymousAccess)
+    }
+    catch (Exception e) {
+      def message = 'Error creating RepositoryManagerClient'
+      logger.println(message)
+      logger.println('Failing build due to error creating RepositoryManagerClient')
+      run.setResult(Result.FAILURE)
+      throw e
```

Same question as earlier, would it be better to throw a new exception with your message having the original exception as the cause?